### ISSUE TITLE
Exception when appending to a global variable's list

### ISSUE THREAD

**Issue by smpanaro** [realm/realm-cocoa#3450]
## Goals

I want to use a global object (`globalB`) to initialize another class (`C`). `globalB` has a `List` property with one element appended.

I was following method 2 in #1109 when I came across this. I don't believe it's related to using a wrapper option class, but that is my goal - so if there's another way to achieve this that'd be fine too.

## Expected Results

`C` has a property whose value is `globalB`.

## Actual Results

I get a crash with this exception:
```
*** Terminating app due to uncaught exception 'RLMException', reason: 'Object type '(null)' does not match RLMArray type 'WrappedA'.'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010233ed85 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001040e2deb objc_exception_throw + 48
	2   Realm                               0x0000000101ddd14d _ZL29RLMValidateMatchingObjectTypeP8RLMArrayP9RLMObject + 328
	3   Realm                               0x0000000101ddd381 -[RLMArray insertObject:atIndex:] + 44
	4   Realm                               0x0000000101ddc9ca -[RLMArray addObject:] + 60
	5   RealmSwift                          0x0000000101d22ecc _TFC10RealmSwift4List6appendfxT_ + 76
	6   test-proj-2                         0x0000000101c72ea1 _TFIv11test_proj_27globalBCS_1BiU_FT_S0_ + 97
	7   test-proj-2                         0x0000000101c71d69 globalinit_33_77D248E752EA58DB4984A4112758C56F_func0 + 9
	8   libdispatch.dylib                   0x00000001058fa3eb _dispatch_client_callout + 8
	9   libdispatch.dylib                   0x00000001058e5d68 dispatch_once_f + 543
	10  test-proj-2                         0x0000000101c71da4 _TF11test_proj_2au7globalBCS_1B + 36
	11  test-proj-2                         0x0000000101c72045 _TFC11test_proj_21CcfT_S0_ + 21
	12  test-proj-2                         0x0000000101c72091 _TToFC11test_proj_21CcfT_S0_ + 17
	13  Realm                               0x0000000101de947f +[RLMObjectSchema propertiesForClass:isSwift:] + 147
	14  Realm                               0x0000000101de879f +[RLMObjectSchema schemaForObjectClass:] + 424
	15  Realm                               0x0000000101e64c3f _ZL16RLMRegisterClassP10objc_class + 133
	16  Realm                               0x0000000101e653d3 __25+[RLMSchema sharedSchema]_block_invoke + 12
	17  CoreFoundation                      0x00000001022701e6 __65-[__NSDictionaryM enumerateKeysAndObjectsWithOptions:usingBlock:]_block_invoke + 102
	18  CoreFoundation                      0x00000001022700da -[__NSDictionaryM enumerateKeysAndObjectsWithOptions:usingBlock:] + 202
	19  Realm                               0x0000000101e6525e +[RLMSchema sharedSchema] + 142
	20  Realm                               0x0000000101e5ae70 +[RLMRealm realmWithConfiguration:error:] + 1575
	21  RealmSwift                          0x0000000101d35af4 _TFC10RealmSwift5RealmCfzT13configurationVS0_13Configuration_S0_ + 820
	22  test-proj-2                         0x0000000101c72d56 _TFIvC11test_proj_214ViewController2DsGC10RealmSwift7ResultsCS_1D_iU_FT_GS2_S3__ + 342
	23  test-proj-2                         0x0000000101c70edd _TFC11test_proj_214ViewControllercfT5coderCSo7NSCoder_GSqS0__ + 29
	24  test-proj-2                         0x0000000101c70fad _TToFC11test_proj_214ViewControllercfT5coderCSo7NSCoder_GSqS0__ + 45
	25  UIKit                               0x0000000102f30095 -[UIClassSwapper initWithCoder:] + 241
	26  UIKit                               0x0000000103107e8a UINibDecoderDecodeObjectForValue + 705
	27  UIKit                               0x0000000103107bc0 -[UINibDecoder decodeObjectForKey:] + 278
	28  UIKit                               0x0000000102f2fd6b -[UIRuntimeConnection initWithCoder:] + 180
	29  UIKit                               0x0000000103107e8a UINibDecoderDecodeObjectForValue + 705
	30  UIKit                               0x000000010310804b UINibDecoderDecodeObjectForValue + 1154
	31  UIKit                               0x0000000103107bc0 -[UINibDecoder decodeObjectForKey:] + 278
	32  UIKit                               0x0000000102f2ef7d -[UINib instantiateWithOwner:options:] + 1255
	33  UIKit                               0x00000001032a9314 -[UIStoryboard instantiateViewControllerWithIdentifier:] + 181
	34  UIKit                               0x00000001032a9467 -[UIStoryboard instantiateInitialViewController] + 69
	35  UIKit                               0x0000000102b6689f -[UIApplication _loadMainStoryboardFileNamed:bundle:] + 94
	36  UIKit                               0x0000000102b66bcf -[UIApplication _loadMainInterfaceFile] + 260
	37  UIKit                               0x0000000102b653ef -[UIApplication _runWithMainScene:transitionContext:completion:] + 1392
	38  UIKit                               0x0000000102b62714 -[UIApplication workspaceDidEndTransaction:] + 188
	39  FrontBoardServices                  0x00000001073838c8 __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 24
	40  FrontBoardServices                  0x0000000107383741 -[FBSSerialQueue _performNext] + 178
	41  FrontBoardServices                  0x0000000107383aca -[FBSSerialQueue _performNextFromRunLoopSource] + 45
	42  CoreFoundation                      0x0000000102264301 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	43  CoreFoundation                      0x000000010225a22c __CFRunLoopDoSources0 + 556
	44  CoreFoundation                      0x00000001022596e3 __CFRunLoopRun + 867
	45  CoreFoundation                      0x00000001022590f8 CFRunLoopRunSpecific + 488
	46  UIKit                               0x0000000102b61f21 -[UIApplication _run] + 402
	47  UIKit                               0x0000000102b66f09 UIApplicationMain + 171
	48  test-proj-2                         0x0000000101c73422 main + 114
	49  libdyld.dylib                       0x000000010592e92d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

## Steps to Reproduce

I included an XCode project that reproduces this crash on my machine.

One interesting thing I noticed is that changing the class names seems to fix it. I've included a commented out section of code that illustrates that.

I also included a commented out section of code that highlights what I want to ultimately be able to do, but it includes a property that isn't necessary to reproduce the crash.

## Code Sample

I've included an XCode project that reproduces it: [test-proj-2.zip](https://github.com/realm/realm-cocoa/files/225278/test-proj-2.zip) You'll need to build the Carthage dependencies.

It's a bare project with this in ViewController.swift.
```swift
import UIKit
import RealmSwift

class ViewController: UIViewController {

    // Won't trigger without accessing Realm.
    var Ds: Results<D> = {
        return try! Realm().objects(D)
    }()
}

// Minimal amount of code to reproduce (as far as I can tell)
class A: Object {}
class WrappedA: Object {}

class B: Object {
    let wrappedAs = List<WrappedA>()
}

let globalB: B = {
    let b = B()
    let w = WrappedA()
    b.wrappedAs.append(w)
    return b
}()

class C: Object {
    dynamic var b: B? = globalB
}

class D: Object {}

// What I want to do ultimately. Similar to method 2 in #1109
/*
class A: Object {}
class WrappedA: Object {
    dynamic var actualA: A?
}

class B: Object {
    let wrappedAs = List<WrappedA>()
}

let globalB: B = {
    let b = B()
    let w = WrappedA()
    w.actualA = A() // Still crashes even without this line.
    b.wrappedAs.append(w)
    return b
}()

class C: Object {
    dynamic var b: B? = globalB
}

class D: Object {}
*/

// Potentially useful:
// Same code as first reproduction except classes A and WrappedA are swapped.
// Does not crash.
/*
class WrappedA: Object {}
class A: Object {
    dynamic var actualA: WrappedA?
}

class B: Object {
    let wrappedAs = List<A>()
}

let globalB: B = {
    let b = B()
    let w = A()
    w.actualA = WrappedA() // Still crashes even without this line.
    b.wrappedAs.append(w)
    return b
}()

class C: Object {
    dynamic var b: B? = globalB
}

class D: Object {}
*/
```

## Version of Realm and Tooling

```
ProductName:	Mac OS X
ProductVersion:	10.11.4
BuildVersion:	15E65

/Applications/Xcode.app/Contents/Developer
Xcode 7.3
Build version 7D175

~/.rvm/gems/ruby-2.1.2/bin/pod
0.36.0
(not in use here)

/bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin15)

/usr/local/bin/carthage
0.11.0
github "realm/realm-cocoa" "v0.98.8"

/usr/bin/git
git version 2.6.4 (Apple Git-63)
```

**Comment by mrackwitz** [comment]
What you want to achieve here is currently not possible with the current design philosophy of the Realm bindings.

Realm scans all classes available in the runtime for subclasses of `Realm.Object` / `RLMObject`. The reflected schema is read, when you first access a Realm database. Swift classes need to be instantiated to fully process all their property types due to limitations of the language. But when `C` is initialized, this triggers the evaluation of `globalB`. `globalB` initializes further types, which might or might not yet be known to Realm. In your case `WrappedA` is not known yet seemingly. There is no way, we could possibly recognize such dependencies automatically.

#### TL;DR

Don't access Realm runtime objects when using the default initialization strategy for objects. (This includes the default values and the default constructor, the one without any arguments.) Recursive schema access could cause errors or dead-locks.

**Comment by smpanaro** [comment]
Ah, I think I understand.

So even something like:
```swift
class B: Object {}
class C: Object {
    dynamic var b: B? = B()
}
```
isn't safe because there's no guarantee that Realm knows about `B` before scanning `C`. Scanning `C` would initialize it which would in turn initialize an instance of `B`. Is that right?

Thanks for the detailed explanation, it's much appreciated :)
Realm is great btw

**Comment by mrackwitz** [comment]
In general that's the idea. But that particular example would work, because there is no schema access necessary to initialize RealmSwift's `Object` subclasses. `List` are different, because even though they are type-safe with Swift, they are a wrapper around `RLMArray`, where you could theoretically try to append objects of others types and it would compile, but our runtime checks make sure to reject those as links have to point always to the same type. Those runtime checks need access to the fully initialized schema.
I think the error message you get in that case, could be more helpful, for that reason, I'll promote this ticket to `T:Enhancement`.

**Comment by mrackwitz** [comment]
Overdue update here: I changed my mind and created a new issue with a summary of the problem here #3454, which was fixed by #3461. This was released with 0.99.
